tests/monotouch-test/Network/NWBrowserTest.cs

```diff
@@ -78,14 +80,70 @@ public void TestStart ()
 		[Test]
 		public void TestStateChangesHandler ()
 		{
-			var e = new AutoResetEvent (false);
-			browser.SetStateChangesHandler ((st, er) => {
-				Assert.IsNotNull (st, "State");
-				Assert.IsNull (er, "Error");
-				e.Set ();
-			});
-			browser.Start ();
-			e.WaitOne ();
+			bool firstRun = true;
+			bool eventsDone = false;
+			bool listeningDone = false;
+			var changesEvent = new AutoResetEvent (false);
+			var browserReady = new AutoResetEvent (false);
+			var finalEvent = new AutoResetEvent (false);
+			TestRuntime.RunAsync (DateTime.Now.AddSeconds (30), async () => {
+				// start the browser, before the listener
+				browser.SetStateChangesHandler ((st, er) => {
+					Assert.IsNotNull (st, "State");
+					Assert.IsNull (er, "Error");
+					if (st == NWBrowserState.Ready)
+						browserReady.Set ();
+				});
+				browser.SetChangesHandler ((oldResult, newResult) => {
+					// first time, listener appears, so we do not have an old result, second time
+					// listener goes, so we do not have a new result
+					if (firstRun) {
+						Assert.IsNull (oldResult, "oldResult first run.");
+						Assert.IsNotNull (newResult, "newResult first run");
+						firstRun = false;
+					} else {
+						Assert.IsNotNull (oldResult, "oldResult first run.");
+						Assert.IsNull (newResult, "newResult first run");
+					}
+					changesEvent.Set ();
+					eventsDone = true;
+
+				});
+				browser.Start ();
+				browserReady.WaitOne (30000);
+				// we are going to start a browser, which will not pick any results, because
+				// we have not listeners, in another thread, the listener will be started and
+				// changes will get the changes for a new service
```

Comment:
Maybe it's my lack of NW knowledge, but I'm not sure what's going on in this comment or the following code 🤔

Reply:
The NW API has different classes to interact with available services. With NWListener, they have chosen a name I would personally not have. We could translate NWLister to NWServer. A NWListener is an object/class that is listening to connection which we can interact with using clients, and that is the reason for the ugly name.

The NWBrowser on the hand is a class that allows to browse/find services (listeners) that are exposing a service over Bonjour. So a browser could look in our network for services of a type (titactoe.tcp.) and on a domain ("mycooldomain").

In the test we are doing the following:

1. Start a browser. At this point, we have no listeners (unless someone is exposing it in the lab) and therefore the browser cannot find any services/listeners.
2. Start a listener that is using the same type/domain pair that the browser expects. 
3. Browser picks up the new listener, and sends an event (service found).
4. Listener stops, and the service disappears.
5. The browser is not yet canceled, so it picks up that the service/listener is not longer then and returns it.

My comment is a little cryptic because it expects knowledge on all that, sorry. I'll update it with the above explanation.